compiler/fir/checkers/checkers.jvm/src/org/jetbrains/kotlin/fir/analysis/jvm/checkers/expression/FirJavaGenericVarianceViolationTypeChecker.kt

```diff
@@ -0,0 +1,110 @@
+import org.jetbrains.kotlin.fir.declarations.FirDeclarationOrigin
+import org.jetbrains.kotlin.fir.declarations.FirFunction
+import org.jetbrains.kotlin.fir.expressions.FirFunctionCall
+import org.jetbrains.kotlin.fir.expressions.impl.FirArgumentListWithResolutionInfo
+import org.jetbrains.kotlin.fir.expressions.toResolvedCallableSymbol
+import org.jetbrains.kotlin.fir.originalOrSelf
+import org.jetbrains.kotlin.fir.resolve.inference.inferenceComponents
+import org.jetbrains.kotlin.fir.resolve.substitution.substitutorByMap
+import org.jetbrains.kotlin.fir.symbols.impl.FirTypeParameterSymbol
+import org.jetbrains.kotlin.fir.types.*
+import org.jetbrains.kotlin.types.AbstractTypeChecker
+import org.jetbrains.kotlin.types.model.TypeConstructorMarker
+import org.jetbrains.kotlin.types.model.typeConstructor
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+
+/**
+ * Checks compatibility of variance of type argument for Java collections.
+ *
+ * Java collection interfaces all include methods mutating the collection. Hence, they naturally map to mutable versions of Kotlin
+ * collections. But Kotlin provides immutable collections and it's good practice to enforce immutability when possible. Hence, to make it
+ * easier to use immutable collections in Kotlin, Java collection types are instead mapped to flexible type with mutable collection as the
+ * lower type and immutable collection as upper type. However, flexible types make type checking unsound. Hence we have this check that
+ * catches a common mistake allowed by flexible type.
+ *
+ * Consider a Java method accepting type `List<Object>`, which is mapped to `MutableList<Any?>..List<(out) Any?>`. If one passes a mutable
+ * list of some more concrete type than `Any` (ArrayList<String>, LinkedList<Int>, etc.) to this Java method, then any writes to this
+ * mutable collection could cause `ClassCastException`s if the same mutable list is read elsewhere. It's the purpose of this checker to
+ * reject such code.
+ *
+ * On the other hand, if one passes an immutable list of some more concrete type, then any writes to it on the Java side would cause
+ * `UnsupportedOperationException`, which is expected and the price we pay in order to make immutable collection easier to use. This checker
+ * doesn't do anything to prevent this from happening.
+ */
+object FirJavaGenericVarianceViolationTypeChecker : FirFunctionCallChecker() {
+    private val javaOrigin = setOf(FirDeclarationOrigin.Java, FirDeclarationOrigin.Enhancement)
+
+    override fun check(expression: FirFunctionCall, context: CheckerContext, reporter: DiagnosticReporter) {
+        val calleeFunction = expression.calleeReference.toResolvedCallableSymbol()?.fir as? FirFunction<*> ?: return
+        if (calleeFunction.originalOrSelf().origin !in javaOrigin) {
+            return
+        }
+        val argList = expression.argumentList as? FirArgumentListWithResolutionInfo ?: return
```

Comment:
This new interface seems to be redundant now
One might use `org.jetbrains.kotlin.fir.expressions.FirExpressionUtilKt#getResolvedArgumentMapping` and `FirResolvedArgumentList`

Reply:
Thanks! Sounds good.